### PR TITLE
Scenario objectives

### PR DESCRIPTION
**Pull request by NickAragua**
Phase I of the scenario objective mechanism. This is the base of the functionality that will allow people to define objectives for scenarios, and have them become associated with various effects on the campaign. 

For objectives with units, we do our best to track which units qualify for which objective, but it's not always possible so we give the user the opportunity to override individual units, or the objective, if the objective doesn't have associated units.

A summary of the user-facing changes:
- AtB scenarios have been updated with appropriate objectives
- Replaced "allies" part of scenario resolution with objective resolution
- Replaced hard-coded scenario post-processing with objective processing
- AtB scenarios now display more explicit and detailed objectives in the briefing screen
- Fixed obnoxious layout issue leading to huge vertical scroll bar in scenario resolution UI
- Finalize button no longer mistakenly available for standard AtB scenarios

image

image

### CODE REVIEW

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/AceDuelBuiltInScenario.java [hunk 1]
```diff
@@ -102,4 +105,15 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
 
 		addBotForce(getEnemyBotForce(getContract(campaign), enemyStart, getSpecMissionEnemies().get(0)));
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/AceDuelBuiltInScenario.java]: Whitespace is a little off from here.

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/AlliedTraitorsBuiltInScenario.java [hunk 2]
```diff
@@ -55,4 +58,14 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
 		addBotForce(
 				new BotForce(getContract(campaign).getAllyBotName(), 2, enemyStart, getSpecMissionEnemies().get(0)));
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/AlliedTraitorsBuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/AllyRescueBuiltInScenario.java [hunk 3]
```diff
@@ -97,4 +103,29 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
 		
 		addBotForce(getEnemyBotForce(getContract(campaign), Board.START_N, enemyEntities));
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/AllyRescueBuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/AmbushBuiltInScenario.java [hunk 4]
```diff
@@ -67,4 +70,14 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
 
 		addBotForce(getEnemyBotForce(getContract(campaign), enemyStart, getSpecMissionEnemies().get(0)));
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/AmbushBuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/BaseAttackBuiltInScenario.java [hunk 5]
```diff
@@ -114,7 +124,52 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
         addEnemyForce(secondBotEntities, getLance(campaign).getWeightClass(campaign), campaign);
         BotForce secondBotForce = getEnemyBotForce(getContract(campaign), isAttacker() ? enemyStart : secondAttackerForceStart, 
                 isAttacker() ? getEnemyHome() : secondAttackerForceStart, secondBotEntities);
-        secondBotForce.setName(secondBotForce.getName() + " Force #2");
+        secondBotForce.setName(String.format("%s%s", secondBotForce.getName(), SECOND_ENEMY_FORCE_SUFFIX));
         addBotForce(secondBotForce);
 	}
+	
+	@Override
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/BaseAttackBuiltInScenario.java]: ws seems off

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/BreakthroughBuiltInScenario.java [hunk 6]
```diff
@@ -96,4 +104,20 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
 	public boolean canAddDropShips() {
 		return !isAttacker() && (Compute.d6() == 1);
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/BreakthroughBuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/ChaseBuiltInScenario.java [hunk 7]
```diff
@@ -122,4 +126,22 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
 			en.setDeployRound(Math.max(0, 12 - speed));
 		}
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/ChaseBuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/CivilianHelpBuiltInScenario.java [hunk 8]
```diff
@@ -69,6 +76,27 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
 			getSurvivalBonusIds().add(UUID.fromString(e.getExternalIdAsString()));
 		}
 
-		addBotForce(new BotForce("Civilians", 1, getStart(), otherForce));
+		addBotForce(new BotForce(CIVILIAN_FORCE_ID, 1, getStart(), otherForce));
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/CivilianHelpBuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/CivilianRiotBuiltInScenario.java [hunk 9]
```diff
@@ -74,7 +85,9 @@ public boolean canDeploy(Unit unit, Campaign campaign) {
 	@Override
 	public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntities,
 			ArrayList<Entity> enemyEntities) {
-        setStart(Board.START_S);
+	    // notrh, south, east, west
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/CivilianRiotBuiltInScenario.java]:
```suggestion
	    // north, south, east, west
```

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/CivilianRiotBuiltInScenario.java [hunk 10]
```diff
@@ -107,6 +120,29 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
                     campaign));
         }
         
-        addBotForce(new BotForce("Rebels", 2, Board.START_N, enemyEntities));
+        addBotForce(new BotForce(REBEL_FORCE_ID, 2, AtBDynamicScenarioFactory.getOppositeEdge(boardEdge), enemyEntities));
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/CivilianRiotBuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/ConvoyAttackBuiltInScenario.java [hunk 11]
```diff
@@ -87,4 +92,15 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
 
 		addBotForce(getEnemyBotForce(getContract(campaign), Board.START_CENTER, enemyEntities));
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/ConvoyAttackBuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/ConvoyRescueBuiltInScenario.java [hunk 12]
```diff
@@ -95,4 +103,25 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
         
         addBotForce(getEnemyBotForce(getContract(campaign), Board.START_S, enemyEntities));
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/ConvoyRescueBuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/ExtractionBuiltInScenario.java [hunk 13]
```diff
@@ -103,4 +111,45 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
 			e.printStackTrace();
 		}
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/ExtractionBuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/HideAndSeekBuiltInScenario.java [hunk 14]
```diff
@@ -88,4 +91,26 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
 
 		addBotForce(getEnemyBotForce(getContract(campaign), enemyStart, getEnemyHome(), enemyEntities));
 	}
+	
+    @Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/HideAndSeekBuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/HideAndSeekBuiltInScenario.java [hunk 15]
```diff
@@ -88,4 +91,26 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
...
+        super.setObjectives(campaign, contract);
+        
+        ScenarioObjective destroyHostiles = CommonObjectiveFactory.getDestroyEnemies(contract, 33);
+        ScenarioObjective keepFriendliesAlive = CommonObjectiveFactory.getKeepFriendliesAlive(campaign, contract, this, 50, false); 
+        ScenarioObjective keepAttachedUnitsAlive = CommonObjectiveFactory.getKeepAttachedGroundUnitsAlive(contract, this);
+        
+        // attacker must destroy 50%
+        if(!isAttacker()) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/HideAndSeekBuiltInScenario.java]: is this test backwards based on the comment?

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/HoldTheLineBuiltInScenario.java [hunk 16]
```diff
@@ -75,4 +78,26 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
 	public boolean canAddDropShips() {
 		return isAttacker() && (Compute.d6() == 1);
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/HoldTheLineBuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/HoldTheLineBuiltInScenario.java [hunk 17]
```diff
@@ -75,4 +78,26 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
...
+        super.setObjectives(campaign, contract);
+        
+        ScenarioObjective destroyHostiles = CommonObjectiveFactory.getDestroyEnemies(contract, 33);
+        ScenarioObjective keepFriendliesAlive = CommonObjectiveFactory.getKeepFriendliesAlive(campaign, contract, this, 50, false); 
+        ScenarioObjective keepAttachedUnitsAlive = CommonObjectiveFactory.getKeepAttachedGroundUnitsAlive(contract, this);
+        
+        // attacker must destroy 50%
+        if(!isAttacker()) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/HoldTheLineBuiltInScenario.java]: Is this test backwards based on the comment?

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/OfficerDualBuiltInScenario.java [hunk 18]
```diff
@@ -100,4 +103,15 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
 
 		addBotForce(getEnemyBotForce(getContract(campaign), enemyStart, getSpecMissionEnemies().get(0)));
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/OfficerDualBuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/PirateFreeForAllBuiltInScenario.java [hunk 19]
```diff
@@ -85,6 +90,19 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
                             campaign));
         }
         
-        addBotForce(new BotForce("Pirates", 3, Board.START_S, otherForce));
+        addBotForce(new BotForce(PIRATE_FORCE_ID, 3, Board.START_S, otherForce));
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/PirateFreeForAllBuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/PrisonBreakBuiltInScenario.java [hunk 20]
```diff
@@ -82,6 +91,27 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
 			getSurvivalBonusIds().add(UUID.fromString(e.getExternalIdAsString()));
 		}
 		
-		addBotForce(new BotForce("Prisoners", 1, getStart(), otherForce));
+		addBotForce(new BotForce(PRISONER_FORCE_ID, 1, getStart(), otherForce));
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/PrisonBreakBuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/PrisonBreakBuiltInScenario.java [hunk 21]
```diff
@@ -82,6 +91,27 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
...
+        super.setObjectives(campaign, contract);
+        
+        //ScenarioObjective destroyHostiles = CommonObjectiveFactory.getDestroyEnemies(contract, 66);
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/PrisonBreakBuiltInScenario.java]: Is this still needed?

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/ProbeBuiltInScenario.java [hunk 22]
```diff
@@ -58,4 +61,20 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
 
 		addBotForce(getEnemyBotForce(getContract(campaign), enemyStart, getEnemyHome(), enemyEntities));
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/ProbeBuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/ReconRaidBuiltInScenario.java [hunk 23]
```diff
@@ -70,4 +78,42 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
 	public boolean canAddDropShips() {
 		return isAttacker() && (Compute.d6() <= 3);
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/ReconRaidBuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/StandUpBuiltInScenario.java [hunk 24]
```diff
@@ -51,6 +54,22 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
...
+	    return Compute.d6() <= 2;
+	}
+	
+	@Override
+	public void setObjectives(Campaign campaign, AtBContract contract) {
+	    super.setObjectives(campaign, contract);
+	    
+	    ScenarioObjective destroyHostiles = CommonObjectiveFactory.getDestroyEnemies(contract, 50);
+	    ScenarioObjective keepFriendliesAlive = CommonObjectiveFactory.getKeepFriendliesAlive(campaign, contract, this, 50, false); 
+        ScenarioObjective keepAttachedUnitsAlive = CommonObjectiveFactory.getKeepAttachedGroundUnitsAlive(contract, this);
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/StandUpBuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/StarLeagueCache1BuiltInScenario.java [hunk 25]
```diff
@@ -112,9 +117,22 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
...
-		loot.setName("Star League Mek");
+		loot.setName(defaultResourceBundle.getString("battleDetails.starLeagueCache.Mek"));
 		loot.addUnit(en);
 		getLoot().add(loot);
-		addBotForce(new BotForce("Tech", 1, getStart(), otherForce));
+		addBotForce(new BotForce(TECH_FORCE_ID, 1, getStart(), otherForce));
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/StarLeagueCache1BuiltInScenario.java]: ws

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/StarLeagueCache2BuiltInScenario.java [hunk 26]
```diff
@@ -55,4 +58,15 @@ public void setExtraMissionForces(Campaign campaign, ArrayList<Entity> allyEntit
 
 		addBotForce(getEnemyBotForce(getContract(campaign), enemyStart, getSpecMissionEnemies().get(0)));
 	}
+	
+	@Override
+    public void setObjectives(Campaign campaign, AtBContract contract) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/StarLeagueCache2BuiltInScenario.java]: ws

File: MekHQ/src/mekhq/gui/dialog/ObjectiveEditPanel.java [hunk 27]
```diff
@@ -0,0 +1,83 @@
+package mekhq.gui.dialog;
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/gui/dialog/ObjectiveEditPanel.java]: Copyright header.

File: MekHQ/src/mekhq/gui/dialog/ObjectiveEditPanel.java [hunk 28]
```diff
@@ -0,0 +1,83 @@
...
+    private List<String> getAvailableForceNames() {
+        List<String> retVal = new ArrayList<>();
+        
+        if(currentScenario instanceof AtBScenario) {
+            for(int x = 0; x < ((AtBScenario) currentScenario).getNumBots(); x++) {
+                retVal.add(((AtBScenario) currentScenario).getBotForce(x).getName());
+            }
+        }
+        
+        //for()
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/gui/dialog/ObjectiveEditPanel.java]: to be continued?

File: MekHQ/src/mekhq/gui/dialog/ResolveScenarioWizardDialog.java [hunk 29]
```diff
@@ -1066,8 +1030,83 @@ private void initComponents() {
...
+                for(String unitID : potentialObjectiveUnits.get(objective)) {
+                    UUID uuid = UUID.fromString(unitID);
+                    if(!tracker.getAllInvolvedUnits().containsKey(uuid)) {
+                        continue;
+                    }
+                    
+                    JCheckBox chkItemState = new JCheckBox(tracker.getAllInvolvedUnits().get(uuid).getShortName());
+                    chkItemState.setSelected(qualifyingObjectiveUnits.get(objective).contains(unitID));
+                    chkItemState.setActionCommand(unitID);
+                    chkItemState.addItemListener(new ItemListener() {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/gui/dialog/ResolveScenarioWizardDialog.java]: Perhaps just:
```java
chkItemState.addItemListener(e -> updateObjectiveDisplay(objective, lblObjective));
```

File: MekHQ/src/mekhq/gui/dialog/ResolveScenarioWizardDialog.java [hunk 30]
```diff
@@ -1546,6 +1694,30 @@ private void showPrisoner(UUID id) {
...
+        
+        for(JCheckBox checkBox : objectiveCheckboxes.get(objective)) {
+            if(checkBox.isSelected()) {
+                qualifyingUnitCount++;
+            }
+        }
+        
+        boolean objectiveMet = objectiveProcessor.objectiveMet(objective, qualifyingUnitCount);
+        if(objectiveMet) {
+            label.setForeground(Color.green.darker());
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/gui/dialog/ResolveScenarioWizardDialog.java]: That's a new one for me; nice.

File: MekHQ/src/mekhq/gui/utilities/Java2sAutoComboBox.java [hunk 31]
```diff
@@ -0,0 +1,82 @@
+package mekhq.gui.utilities;
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/gui/utilities/Java2sAutoComboBox.java]: Copyright header.

File: MekHQ/src/mekhq/gui/dialog/ObjectiveEditPanel.java (hunk 28 above)
Comment (NickAragua) [reply MekHQ/src/mekhq/gui/dialog/ObjectiveEditPanel.java]: Yeah, this file is incomplete. I plan to include an objective editor feature in the next PR.

File: MekHQ/src/mekhq/gui/utilities/Java2sAutoComboBox.java (hunk 31 above)
Comment (NickAragua) [reply MekHQ/src/mekhq/gui/utilities/Java2sAutoComboBox.java]: This one is weird, as I lifted it straight from stack overflow without even pretending to modify it. Not sure how that should go re: copyright header.

Comment (NickAragua) [reply MekHQ/src/mekhq/gui/utilities/Java2sAutoComboBox.java]: Yeah, turns out I'll be removing this - it doesn't really do what I need it to do and my implementation of the relevant code went in a different direction.

File: MekHQ/src/mekhq/campaign/mission/atb/scenario/HoldTheLineBuiltInScenario.java [hunk 32]
```diff
@@ -15,88 +34,91 @@
...
+        ScenarioObjective keepFriendliesAlive = CommonObjectiveFactory.getKeepFriendliesAlive(campaign, contract, this,
+                50, false);
+        ScenarioObjective keepAttachedUnitsAlive = CommonObjectiveFactory.getKeepAttachedGroundUnitsAlive(contract,
+                this);
+
+        // defender must destroy 50%
+        // attacker must destroy 33%
+        if (!isAttacker()) {
             destroyHostiles.setPercentage(50);
             keepFriendliesAlive.setPercentage(33);
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/atb/scenario/HoldTheLineBuiltInScenario.java]: If the attackers must destroy 33%, wouldn't you have to keep 66% alive?

Comment (NickAragua) [reply MekHQ/src/mekhq/campaign/mission/atb/scenario/HoldTheLineBuiltInScenario.java]: Good point.

Comment (NickAragua) [reply MekHQ/src/mekhq/campaign/mission/atb/scenario/HoldTheLineBuiltInScenario.java]: I'm going to address it in a separate PR, as I already made a bunch of commits on top of the branch I used to develop this and it's kind of a pain to separate them out.

**Pull request merged**